### ISSUE TITLE
Add moment.js in the Postman sandbox

### ISSUE THREAD

**Issue by amirabiri** [postmanlabs/postman-app-support#553]
The timestamp dynamic variable is very useful for me. However I also need an equivalent in ISO 8601 format, i.e:

```
POST /.../
{
    ...
    "actionTime": "2014-04-19T 17:05:53Z"
}
```

I think there should be two variables: one for local time, one for UTC.

**Comment by ssi-anik** [comment]
So, If  I want to send a time dynamically to my backend, I've to set an environment variable, pass it to the body? and then to backend?

**Comment by madebysid** [comment]
@ssi-anik You can generate a timestamp in your pre-request script, and set it as an environment variable like @abhijitkane illustrated above. 
You can then use it in the same request's body.

**Comment by sdnts** [comment]
@ssi-anik We have a dynamic variable called $timestamp that can generate a timestamp for you. [Docs](https://www.getpostman.com/docs/sandbox). You can make use of this.

**Comment by ssi-anik** [comment]
I wanted a string like '2016-09-29 11:25 PM', So, I used the Date. Saved on a variable, passed that, that's it. 
Thanks anyway. :)

**Comment by emersonf** [comment]
Any chance this can get revisited? Writing a pre-request script for every request that needs an ISO8601 timestamp creates a lot of repetition, especially if any customization is required, e.g. to get the local date or whatever.

I assume two timestamp formats cover most cases, namely a local timestamp `{{$timestampIso8601Local}}`, e.g. `2017-05-28T22:12:01+01:00` and UTC time `{{$timestampIso8601Utc}}`, e.g. `2017-05-28T21:12:01Z`.

A thorough set of time-based dynamic variables would make people even happier though.

**Comment by lsloan** [comment]
I arrived here looking for an answer to the question:  How can I get the current timestamp for UTC in ISO 8601 format to appear in the body data of a Postman request?

In case anyone else has arrived here with the same question, this is the solution (using Postman v5.0.1 for Chrome 58.0.3029.110 on macOS 10.12.5) I used:

1. In the Builder, while editing a request, click the "Pre-request Script" heading below the URL field.
1. In the editor field that appears, enter this single line of JavaScript:

    ```javascript
    postman.setGlobalVariable('timestampUtcIso8601', (new Date()).toISOString());
    ```
1. Click the "Body" heading below the URL field.
1. In the editor field that appears, wherever you want the newly formatted timestamp to appear, add the reference `{{timestampUtcIso8601}}`.

    > 💡 Remember: Don't use a dollar sign (`$`) in this variable reference.  That's only needed for reserved system variables automatically set by Postman.  E.g., the system timestamp (the number of seconds since 00:00:00 UTC, 1 Jan 1970), `{{$timestamp}}`.

How this works is Postman runs the "Pre-request Script" before each request, which updates the `timestampUtcIso8601` global variable with the current time.  When the request references the variable, it will be replaced with its most recent value.

As others have written here before (@emersonf, who also _[reiterated his concerns](553#issuecomment-310477486) in response to this comment_), it's repetitive and tedious to add this "Pre-request Script" to every request that needs the timestamp.  It would be helpful if Postman [supported scripts at the collection (folder) level](882) (#882).  Maybe even better would be for it to [support JavaScript when setting values for environment and global variables](160) (#160).

Fortunately, it seems that once the variable has been set, it will remain available for other requests that follow it.  They can use the previously-set value without updating it, but it will be getting further in the past.

PS: Note these related issues:

* Issue #882 requests a feature for pre-request scripts at the collection level.  If you like this idea, go there and vote for the issue.  There are also some options for which @shamasis wants feedback, so give your opinion.

* Issue #160 requests a feature to allow environment/global variables to contain JavaScript that gets executed each time they're referenced.

**Comment by emersonf** [comment]
While this works, it requires copy and pasting that JS to every single
request that needs it, which is terribly inconvenient and prone to error.

On 22 Jun 2017 6:36 p.m., "Lance E Sloan «UMich»" <notifications@github.com>
wrote:

I arrived here looking for an answer to the question: How can I get the
current timestamp for UTC in ISO 8601 format to appear in the body data of
a Postman request?

In case anyone else has arrived here with the same question, this is the
solution (using Postman v5.0.1 for Chrome 58.0.3029.110 on macOS 10.12.5) I
used:

   1.

   In the Builder, while editing a request, click the "Pre-request Script"
   heading below the URL field.
   2.

   In the editor field that appears, enter this single line of JavaScript:

   postman.setGlobalVariable("timestampUtcIso8601", (new Date()).toISOString());

   3.

   Click the "Body" heading below the URL field.
   4.

   In the editor field that appears, wherever you want the newly formatted
   timestamp to appear, add the reference {{timestampUtcIso8601}}.

   💡 Remember: Don't use a dollar sign ($) in this variable reference.
   That's only needed for reserved system variables automatically set by
   Postman. E.g., the system timestamp (the number of seconds since 00:00:00
   UTC, 1 Jan 1970), {{$timestamp}}.


How this works is Postman runs the "Pre-request Script" before each
request, which updates the timestampUtcIso8601 global variable with the
current time. When the request references the variable, it will be replaced
with its most recent value.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/postmanlabs/postman-app-support/issues/553#issuecomment-310434775>,
or mute the thread

**Comment by sdnts** [comment]
Guys, we've been working on adding `moment.js` to our sandbox, which should solve all these issues. I'm changing the issue title to reflect this as well.

**Comment by lsloan** [comment]
Thanks for the update, @madebysid.  Having `moment.js` available will be helpful for formatting timestamps.  However, I don't understand how it would help with this issue.

The original issue was that Postman needs "Current timestamp, in ISO 8601 format, dynamic variable".  That is, timestamp system variables, like `$timestamp`, but their values would be in ISO 8601 format.  There should be two such variables, one with UTC time and the other with the local time.

How will adding `moment.js` accomplish this?  As you can see from my recent comment, using standard JavaScript, we can already get the timestamp in that format:

```javascript
(new Date()).toISOString()
```

The **_problem_** that this issue documents is not that we need `moment.js`.  The **_problem_** is that the formatted timestamps should be available as system variables.  In order to get this kind of timestamp in our requests now, we need to add a "Pre-request Script" to each request which sets the timestamp in global or environment variables.

I'm not against the addition of `moment.js` to the Postman sandbox, but I need an explanation of how it addresses this issue.  Maybe it's just the right solution and I don't understand it.  I'd love to be enlightened.

Also, I think you need to _change the title of this issue back_.  Changing the title makes sense if it clarifies the problem.  However, the new title you assigned redefines the whole issue.  It no longer means the same thing as what @amirabiri originally posted.

Do others at Postman Labs (like @shamasis, @abhijitkane, @postman-admin, etc.) have an opinion on this?

**Comment by kunagpal** [comment]
@lsloan @ssi-anik @amirabiri @emersonf Moment.js (without locales) is available in Postman v5.1.2 and above, do go ahead and check it out. You can use `moment` in your scripts as follows:
```js
var moment = require('moment');

moment('2017-01-01T10:10:10.000').format('MMMM Do YYYY, h:mm:ss a') // January 1st 2017, 10:10:10 am
```

**Comment by emersonf** [comment]
@kunagpal @madebysid moment.js is a great library and I'm happy to see it's been added. But adding moment doesn't satisfy what is being asked for here. (as @lsloan said, this issue shouldn't have been renamed)

At face value, we need pre-defined variables for ISO 8601 timestamps. If the suggestion is to use moment to create them in _Pre-request scripts_, then we have a different problem; how do we run the code that creates those variables before *every* request? Copy-pasting is not a viable option.

**Comment by kunagpal** [comment]
@emersonf What you've mentioned is related to https://github.com/postmanlabs/postman-app-support/issues/882

**Comment by srinivas258** [comment]
Hi Could you please let me know how to get delayed timestamp? per1 hour or 2 hour pirior to the sys date?
my request is like below

url?starttime=2017-10-25T12:00&endtime=2017-10-25T13:00 also this is fine(end time = sysdate) rite now i am doing manually want to do it dynamically

http://host.php?timestamp=2017-10-24T13:00&endTimestamp={{timestampUtcIso8601}}

**Comment by Stevoisiak** [comment]
This page is the first result when searching "Postman Format Timestamp" and "Postman Date Format", so this issue seems to be somewhat common.

At the very least I would like to see this issue re-opened, since the issue has yet to be solved.

**Comment by gekonwi** [comment]
Simple solution: add a global pre-request script!

We have a well defined hierarchy of variables: global > environment > collection > local.

And we have pre-request scripts for: collection, folder, and local. We could consistently add pre-request scripts for the two missing levels: global and environment.

Having that, we could define the timestamp as needed in the global pre-request script once, for all requests in the workspace, avoiding the headache of changing the script in dozens of requests if it ever needs to be changed.

@madebysid - what do you think?

**Comment by jpjpjp** [comment]
Old thread.   The pre-request script is OK for a lot of situations but if you are creating Example Responses for a mock server its not an option.   So it still would be great if there was a dynamic variable like $utc_timestamp

**Comment by mohitranka** [comment]
Released in `v 7.29.0`.
